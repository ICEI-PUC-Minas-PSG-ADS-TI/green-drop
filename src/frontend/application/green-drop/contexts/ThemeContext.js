import React, { createContext, useContext, useState, useEffect } from "react";
import { useColorScheme } from "react-native";
import PropTypes from 'prop-types';

export const ThemeContext = createContext(
  {
  colorScheme: "light",
  toggleTheme: () => {},
}
);

export const ThemeProvider = ({ children }) => {
  const systemColorScheme = useColorScheme();
  const [colorScheme, setColorScheme] = useState(systemColorScheme);
  if (__DEV__) {
    console.log("Theme is:", colorScheme);
  }

  useEffect(() => {
    setColorScheme(systemColorScheme);
  }, [systemColorScheme]);

  const toggleTheme = () => {
    setColorScheme(prev => (prev === 'dark' ? 'light' : 'dark'));
  };

  return (
    <ThemeContext.Provider value={{ colorScheme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export const useTheme = () => useContext(ThemeContext);