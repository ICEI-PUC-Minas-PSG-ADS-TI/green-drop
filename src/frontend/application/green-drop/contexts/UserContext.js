import React, { createContext, useContext, useState, useEffect } from "react";
import PropTypes from "prop-types";
import { usuarioService } from '@/services/Usuario';

const EXPIRATION_TIME = 3 * 24 * 60 * 60 * 1000; // 3 dias em milissegundos
export const UserContext = createContext();

export function UserProvider({ children }) {
    const [user, setUser] = useState(null);
    const [loginStatus, setLoginStatus] = useState(false);
    const [loading, setLoading] = useState(false);
    const [conquistas, setConquistas] = useState([]);
    const [desafios, setDesafios] = useState([]);

    useEffect(() => {
        if (__DEV__) simularLoginDev();
    }, []);

    const simularLoginDev = () => {
        setLoginStatus(true);
        const userMock = {
            displayName: "Usuário de Teste",
            email: "teste@gmail.com",
            telefone: "1234567890",
            pontos: 150,
            xp: 50,
        };
        const conquistasMock = [
            { id: 1, titulo: "Primeiros Passos", descricao: "Faça sua primeira contribuição", completo: true },
            { id: 2, titulo: "Explorador Nato", descricao: "Visite 50 locais diferentes", completo: false },
        ];
        const desafiosMock = [
            { id: 1, titulo: "Explorador Iniciante", descricao: "Adicione 5 novos locais", progresso: "3/5", completo: false },
            { id: 2, titulo: "Contribuidor Ativo", descricao: "Faça 10 contribuições este mês", progresso: "7/10", completo: false },
        ];

        setUser(userMock);
        setConquistas(conquistasMock);
        setDesafios(desafiosMock);
    };

    const loginUsuario = async (identificador, senha) => {
        setLoading(true);
        try {
            const isEmail = /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(identificador);
            const isTelefone = /^\+?\d{10,15}$/.test(identificador);

            if (!isEmail && !isTelefone) {
                throw new Error("Identificador inválido. Use um e-mail ou telefone válido.");
            }

            const data = await usuarioService.login({
                email: isEmail ? identificador : undefined,
                telefone: isTelefone ? identificador : undefined,
                senha,
            });

            setUser(data);
            setLoginStatus(true);
            return true;
        } catch (error) {
            console.error("Erro ao fazer login:", error.message || error);
            return false;
        }
        finally {
            setLoading(false);
        }
    };


    const logoutUsuario = () => {
        setUser(null);
        setLoginStatus(false);
    };

    const cadastrarUsuario = async ({ nome, email, senha, telefone }) => {
        setLoading(true);
        try {
            await usuarioService.cadastro({ nome, email, senha, telefone })
            await loginUsuario(email, senha);
        } catch (error) {
            console.error("Erro ao cadastrar usuário:", error.message);
            return false;
        } finally {
            setLoading(false);
        }
    };

    return (
        <UserContext.Provider
            value={{
                user,
                conquistas,
                desafios,
                loginStatus,
                loading,
                loginUsuario,
                logoutUsuario,
                cadastrarUsuario,
            }}
        >
            {children}
        </UserContext.Provider>
    );
}
UserProvider.propTypes = { children: PropTypes.node };
export const useUserContext = () => useContext(UserContext);