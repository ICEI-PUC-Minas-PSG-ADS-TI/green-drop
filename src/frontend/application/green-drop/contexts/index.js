export { ThemeProvider, useTheme } from './ThemeContext'
export { UserProvider, useUserContext } from './UserContext'