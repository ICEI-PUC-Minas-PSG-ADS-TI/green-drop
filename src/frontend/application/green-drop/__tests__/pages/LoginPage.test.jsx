import React from 'react';
import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { NavigationContainer } from '@react-navigation/native';
import { ThemeProvider } from '@/contexts/ThemeContext.js';
import { UserProvider } from '@/contexts/UserContext';
import LoginPage from '../../pages/login/index.jsx';

describe('<LoginPage />', () => {
  it('renderiza corretamente', () => {
    render(
      <SafeAreaProvider>
        <ThemeProvider>
          <UserProvider>
            <NavigationContainer>
              <LoginPage />
            </NavigationContainer>
          </UserProvider>
        </ThemeProvider>
      </SafeAreaProvider>
    );
  });

  it('snapshot de LoginPage', () => {
    const tree = renderer.create(
      <SafeAreaProvider>
        <ThemeProvider>
          <UserProvider>
            <NavigationContainer>
              <LoginPage />
            </NavigationContainer>
          </UserProvider>
        </ThemeProvider>
      </SafeAreaProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
