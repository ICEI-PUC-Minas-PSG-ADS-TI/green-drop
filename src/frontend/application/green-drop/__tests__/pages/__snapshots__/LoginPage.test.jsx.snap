// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> snapshot de LoginPage 1`] = `null`;
